validation: Tidy up ValidationState interface

jnewbery opened the pull request:
Carries out some remaining tidy-ups remaining after PR 15141:

- split ValidationState into TxValidationState and BlockValidationState (commit from ajtowns)
- various minor code style tidy-ups to the ValidationState class
- remove the useless `ret` parameter from `ValidationState::Invalid()`
- remove the now unused `first_invalid` parameter from `ProcessNewBlockHeaders()`
- remove the `fMissingInputs` parameter from `AcceptToMemoryPool()`, and deal with missing inputs the same way as other errors by using the `TxValidationState` object.

Tip for reviewers (thanks ryanofsky!): The first commit ("[validation] Add CValidationState subclasses" ) is huge and can be easier to start reviewing if you revert the rote, mechanical changes:

Substitute the commit hash of commit "[validation] Add CValidationState subclasses" for <CommitHash> in the commands below.

```sh
git checkout <CommitHash>
git grep -l ValidationState | xargs sed -i 's/BlockValidationState\|TxValidationState/CValidationState/g'
git grep -l ValidationResult | xargs sed -i 's/BlockValidationResult\|TxValidationResult/ValidationInvalidReason/g'
git grep -l MaybePunish | xargs sed -i 's/MaybePunishNode\(ForBlock\|ForTx\)/MaybePunishNode/g'
git diff HEAD^
```

After that it's possible to easily see the mechanical changes with:

```sh
git log -p -n1 -U0 --word-diff-regex=. <CommitHash>
```

jnewbery committed:
[validation] Tidy Up ValidationResult class

Minor style fixups and comment updates.

This is purely a style change. There is no change in behavior.

jnewbery committed:
[validation] Remove error() calls from Invalid() calls

This is in preparation for the next commit, which removes the useless
`ret` parameter from ValidationState::Invalid().

error() is simply a convenience wrapper that calls LogPrintf and returns
false. Call LogPrintf explicitly and substitute the error() call for a
false bool literal.

jnewbery committed:
[validation] Remove useless ret parameter from Invalid()

ValidationState::Invalid() takes a parameter `ret` which is returned to
the caller. All call sites set this to false. Remove the `ret` parameter
and just return false always.

jnewbery committed:
[validation] Remove unused first_invalid parameter from ProcessNewBlockHeaders()

No callers use the returned value in first_invalid. Remove it from the
function signature and don't set it in the function.

jnewbery committed:
[validation] Remove fMissingInputs from AcceptToMemoryPool()

Handle this failure in the same way as all other failures: call Invalid()
with the reasons for the failure.

jnewbery commented:
This is built on #15141. Only commit _[validation] Add CValidationState subclasses_ onwards are for review in this PR.

promag reviewed (commented):
Concept ACK, looks like a good refactor and improvement to add ValidationState subclasses.

JeremyRubin commented:
Concept ACK.

I'd like to see `Invalid` and ilk return void instead of always false, and clean up the call sites correspondingly. This adds a  lines here and there, but I think it improves the readability of the code to see a false literal returned explicitly.

I also somewhat think that ideally there should be a third state class which handles Corruption cases for 'system state' or something. This covers the notion that the issue is not the fault of the block or the transaction, it is an issue with our local state or implementation. Not sure it's worth the churn though, two is already a huge improvement.

laanwj commented:
Concept ACK

ajtowns commented:
> I also somewhat think that ideally there should be a third state class which handles Corruption cases for 'system state' or something. This covers the notion that the issue is not the fault of the block or the transaction,

The current split is more about "what was being tested" not "what was at fault" -- that's easy to deal with via types, because you know what's being tested at compile time; but you don't know what's going to have been at fault at compile time, so I think dealing with that via the type system (ie as a third state class) wouldn't work.

practicalswift commented on src/validation.h:
```diff
@@ -228,7 +229,7 @@ bool ProcessNewBlock(const CChainParams& chainparams, const std::shared_ptr<cons
  * @param[out] ppindex If set, the pointer will be set to point to the last new block index object for the given headers
  * @param[out] first_invalid First header that fails validation, if one exists
```
`first_invalid` has been dropped, right? :-)

jnewbery replied:
Good catch. Fixed.

jnewbery commented:
rebased: 8d88c8608

jnewbery commented:
Fixed @practicalswift's comment: b4c8f0079059772f1bb662820c5cce8dab66237d

flack commented on src/consensus/validation.h:
```diff
@@ -12,24 +12,17 @@
-/** "reject" message codes */
-static const unsigned char REJECT_MALFORMED = 0x01;
+/** A reject code must be provided for P2P reject messages. Just always send REJECT_INVALID
+ *  TODO: remove support for BIP 61 reject messages.
```
nit: `*` is not aligned quite right

```suggestion
  * TODO: remove support for BIP 61 reject messages.
```

jnewbery replied:
thanks. Fixed!

jnewbery commented:
rebased

jnewbery commented:
rebased

jnewbery commented:
rebased

ryanofsky commented on src/bench/block_assemble.cpp:
```diff
@@ -39,7 +39,7 @@ static void AssembleBlock(benchmark::State& state)
 
         for (const auto& txr : txs) {
-            CValidationState state;
+            TxValidationState state;
```
In commit "[validation] Add CValidationState subclasses" (1faf0c4b0120995b4361d38ca0cc0114e633955d)

Note to other reviewers: This first commit is huge and can be easier to start reviewing if you revert the rote, mechanical changes:

```sh
git checkout 1faf0c4b0120995b4361d38ca0cc0114e633955d
git grep -l ValidationState | xargs sed -i 's/BlockValidationState\|TxValidationState/CValidationState/g'
git grep -l ValidationResult | xargs sed -i 's/BlockValidationResult\|TxValidationResult/ValidationInvalidReason/g'
git grep -l MaybePunish | xargs sed -i 's/MaybePunishNode\(ForBlock\|ForTx\)/MaybePunishNode/g'
git diff HEAD^
```

After that it's possible to review the full change with:

```sh
git log -p -n1 -U0 --word-diff-regex=. 1faf0c4b0120995b4361d38ca0cc0114e633955d
```

jnewbery replied:
Thanks Russ. I've added that review tip to the PR description.

ryanofsky reviewed (commented):
Started review (will update list below with progress).
- [X] 1faf0c4b0120995b4361d38ca0cc0114e633955d [validation] Add CValidationState subclasses (1/7)
- [x] a3076817588a642b4de107dc96a76ea8939070df [validation] Remove Reject Code from ValidationState (2/7)
- [ ] 00e997c21af1e107c4370a6cae15b19a293cae63 [validation] Tidy Up ValidationResult class (3/7)
- [ ] eba4651d76f8f636953cacfcb0979c10ec1025ea [validation] Remove error() calls from Invalid() calls (4/7)
- [ ] 578a4a166f575174cf38ae55131f96750c807bdf [validation] Remove useless ret parameter from Invalid() (5/7)
- [ ] f21373b0211bd1f175171a2e7592690ff220d008 [validation] Remove unused first_invalid parameter from ProcessNewBlockHeaders() (6/7)
- [ ] 4cef7e521aa1ff7752653bc638e3bfc9a449af26 [validation] Remove fMissingInputs from AcceptToMemoryPool() (7/7)

Note that travis currently seems to be failing.

ajtowns commented on src/validation.cpp:
```diff
@@ -2019,25 +2012,17 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
-                    // In the event of a future soft-fork, we may need to
-                    // consider whether rewriting to CONSENSUS or
-                    // RECENT_CONSENSUS_CHANGE would be more appropriate.
-                    state.Invalid(ValidationInvalidReason::CONSENSUS, false,
```
That looks wrong to me too. Might be clearer to tell `CheckInputs()` which validation rules are consensus vs standardness though?

jnewbery replied:
Yes, this is a bug (and a very bad one!) I'm going to remove the `FromTxValidationState()` method entirely and just set the `BlockValidationState.m_result` through a call to `Invalid()` (see https://github.com/bitcoin/bitcoin/pull/15921#discussion_r289086882).

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -1249,9 +1248,11 @@ void PeerLogicValidation::BlockChecked(const CBlock& block, const BlockValidatio
-        // Don't send reject message with code 0 or an internal reject code.
-        if (it != mapBlockSource.end() && State(it->second.first) && state.GetRejectCode() > 0 && state.GetRejectCode() < REJECT_INTERNAL) {
-            CBlockReject reject = {(unsigned char)state.GetRejectCode(), state.GetRejectReason().substr(0, MAX_REJECT_MESSAGE_LENGTH), hash};
+        // Send reject messages for invalid blocks
```
In commit "[validation] Remove Reject Code from ValidationState" (a3076817588a642b4de107dc96a76ea8939070df)

Can you add a comment about what why BLOCK_MISSING_PREV and BLOCK_CACHED_INVALID are special here? Missing prev case seems to make sense and I see that the cached_invalid case isn't changing behavior, but I don't understand it.

jnewbery replied:
No longer part of this PR.

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -2603,8 +2604,8 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                 FormatStateMessage(state));
-            if (enable_bip61 && state.GetRejectCode() > 0 && state.GetRejectCode() < REJECT_INTERNAL) { // Never send AcceptToMemoryPool's internal codes over P2P
-                connman->PushMessage(pfrom, msgMaker.Make(NetMsgType::REJECT, strCommand, (unsigned char)state.GetRejectCode(),
+            if (enable_bip61 && state.IsInvalid()) {
```
In commit "[validation] Remove Reject Code from ValidationState" (a3076817588a642b4de107dc96a76ea8939070df):

Note: IIUC, there is no change in behavior here because previous `state.GetRejectCode() < REJECT_INTERNAL` condition was always true. REJECT_HIGHFEE is never triggered for p2p transactions.

jnewbery replied:
No longer part of this PR.

ryanofsky commented on src/validation.cpp:
```diff
@@ -588,28 +588,28 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
 
     // Rather not work on nonstandard transactions (unless -testnet/-regtest)
-    std::string reason;
+    std::string reason {};
```
In commit "[validation] Remove Reject Code from ValidationState" (a3076817588a642b4de107dc96a76ea8939070df)

Unintentional change? Doesn't serve an obvious purpose.

jnewbery replied:
No longer part of this PR.

ryanofsky commented on src/validation.cpp:
```diff
@@ -735,22 +735,22 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
         }
 
         if (nAbsurdFee && nFees > nAbsurdFee)
             return state.Invalid(TxValidationResult::TX_NOT_STANDARD, false,
```
In commit "[validation] Remove Reject Code from ValidationState" (a3076817588a642b4de107dc96a76ea8939070df)

Code predates this commit, but TX_MEMPOOL_POLICY would seem to make more sense here than TX_NOT_STANDARD. It might be good to add a comment if using TX_NOT_STANDARD is important here, since it's not the obvious choice.

jnewbery replied:
No longer part of this PR.

jnewbery commented on src/consensus/validation.h:
```diff
@@ -135,12 +133,57 @@ class CValidationState {
+        case TxValidationResult::NONE:
+        case TxValidationResult::TX_NOT_STANDARD:
+        case TxValidationResult::TX_MEMPOOL_POLICY:
+            m_result = BlockValidationResult::NONE;
```
For `FromTxValidationState()` should we assert that `tx_state.GetResult()` does not return `TX_NOT_STANDARD` or `TX_MEMPOOL_POLICY`? This should only be called from block validation code, where transactions shouldn't fail for standardness or policy reasons.

I don't think `Invalid()` can change state from invalid to valid: `m_mode` is either kept as `MODE_ERROR` or changed to `MODE_INVALID`. Can you point out where I'm wrong?

ryanofsky replied:
I'm just saying that this seems error prone, not that there currently is an error. Adding asserts to check the previous state and ensure that calling these methods never changes it from invalid to valid seems like an easy change that could prevent a serious bug. But I could be wrong about this. Maybe it wouldn't be an easy change or wouldn't be helpful in preventing bugs.

jnewbery replied:
In fact, I think I should just remove this method entirely. `FromTxValidationState()` is only called from three places and in each case we want the `BlockValidationState.m_result` to be set to `CONSENSUS`.

jnewbery commented on src/validation.cpp:
```diff
@@ -1437,18 +1436,18 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
                     }
                     // MANDATORY flag failures correspond to
-                    // ValidationInvalidReason::CONSENSUS. Because CONSENSUS
+                    // TxValidationInvalidReason::CONSENSUS. Because CONSENSUS
```
fixed

jnewbery commented on src/validation.cpp:
```diff
@@ -3265,23 +3254,23 @@ static bool ContextualCheckBlockHeader(const CBlockHeader& block, CValidationSta
        (block.nVersion < 3 && nHeight >= consensusParams.BIP66Height) ||
        (block.nVersion < 4 && nHeight >= consensusParams.BIP65Height))
-            return state.Invalid(ValidationInvalidReason::BLOCK_INVALID_HEADER, false, REJECT_OBSOLETE, strprintf("bad-version(0x%08x)", block.nVersion),
+            return state.Invalid(BlockValidationResult::CONSENSUS, false, REJECT_OBSOLETE, strprintf("bad-version(0x%08x)", block.nVersion),
```
No idea. Fixed!

jnewbery commented on src/consensus/validation.h:
```diff
@@ -12,20 +12,13 @@
+/** A reject code must be provided for P2P reject messages. Just always send REJECT_INVALID
+  * TODO: remove support for BIP 61 reject messages.
+  */
 static const unsigned char REJECT_INVALID = 0x10;
```
I'm not sure if that's possible. The BIP61 reject codes are defined here: https://github.com/bitcoin/bips/blob/master/bip-0061.mediawiki#reject_version_codes.

I hope we'll remove REJECT messages soon (https://github.com/bitcoin/bitcoin/pull/15437)

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15921#discussion_r289028051

In commit "[validation] Remove Reject Code from ValidationState" (22be379c54eb03e77803b0bbd8270082452d8fff)

It seems like this commit changing p2p codes would make more sense as a separate PR (maybe based on this PR) not mixed in with all the refactoring changes in this PR. Someone could be indifferent to the refactoring changes but want to raise an issue with the p2p change, or vice versa, so it'd be nice if any discussions would happen in separate places.

Also, I think the text in the commit message should be a release note.

jnewbery replied:
I wanted to get all of these changes into the same PR. All the commits touch the same lines, so splitting the changes across multiple PRs means more rebasing for other PRs. However, what you say makes sense - I shouldn't mix P2P changes with a refactor.

I've removed this commit from the branch, so this PR should be basically a refactor with very minor logging/error message changes.

The removal of Reject Code from the ValidationState object can wait until we remove REJECT messages entirely (https://github.com/bitcoin/bitcoin/pull/15437).

jnewbery commented:
Thanks for the review @ryanofsky and @ajtowns . I've addressed all the feedback on commit _CValidationState subclasses_.

jnewbery commented:
rebased

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -973,38 +973,34 @@ void Misbehaving(NodeId pnode, int howmuch, const std::string& message) EXCLUSIV
  *
  * @return Returns true if the peer was punished (probably disconnected)
- *
- * Changes here may need to be reflected in TxRelayMayResultInDisconnect().
```
In commit "[validation] Add CValidationState subclasses" (b31987c45045cc26ca7867d9d8aa645f5d36b5e6)

Any reason this comment is being deleted? It's little vague, but I think it's just saying that this function bans peers, and that TxRelayMayResultInDisconnect tries to avoid being banned by peers, so a change here might require a change there.

jnewbery replied:
Because this function is to decide whether to punish a peer for a bad block. The function below is to decide whether to punish a peer for a bad tx. Changes to that function need to be reflected in `TxRelayMayResultInDisconnect()` since they're both concerned with txs.

ryanofsky commented on src/validation.cpp:
```diff
@@ -1876,25 +1871,19 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
-                }
+            TxValidationState tx_state;
+            if (!CheckInputs(tx, tx_state, view, fScriptChecks, flags, fCacheResults, fCacheResults, txdata[i], nScriptCheckThreads ? &vChecks : nullptr)) {
+                // Any transaction validation failure in ConectBlock is a block consensus failure
```
In commit "[validation] Add CValidationState subclasses" (b31987c45045cc26ca7867d9d8aa645f5d36b5e6)

spelling ConnectBlock

jnewbery replied:
fixed

ryanofsky commented on src/validation.cpp:
```diff
@@ -1839,20 +1838,16 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
-                }
+            TxValidationState tx_state;
+            if (!Consensus::CheckTxInputs(tx, tx_state, view, pindex->nHeight, txfee)) {
+                // Any transaction validation failure in ConectBlock is a block consensus failure
```
In commit "[validation] Add CValidationState subclasses" (b31987c45045cc26ca7867d9d8aa645f5d36b5e6)

spelling ConnectBlock

jnewbery replied:
fixed

ryanofsky commented on src/validation.cpp:
```diff
@@ -1839,20 +1838,16 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
         {
             CAmount txfee = 0;
-            if (!Consensus::CheckTxInputs(tx, state, view, pindex->nHeight, txfee)) {
-                if (!IsBlockReason(state.GetReason())) {
```
In commit "[validation] Add CValidationState subclasses" (b31987c45045cc26ca7867d9d8aa645f5d36b5e6)

To be sure, is there no change in behavior here because IsBlockReason would always return false if CheckTxInputs set failed with any reason other than CONSENSUS? Or is there no change in behavior for some other reason? Or is there a actually a change in some cases?

Related question: Is it a problem going forward that TxValidationResult::RECENT_CONSENSUS_CHANGE will be translated to BlockValidationResult::CONSENSUS_CHANGE instead of BlockValidationResult::RECENT_CONSENSUS_CHANGE?

jnewbery replied:
Take a look at `CheckTxInputs()`. Prior to this commit, if it returned false, then `state.m_reason` could be set to:

- `TX_MISSING_INPUTS`
- `TX_PREMATURE_SPEND`
- `CONSENSUS`

In all cases, that would be converted to `CONSENSUS`, so this is not a change in behaviour.

> Is it a problem going forward that TxValidationResult::RECENT_CONSENSUS_CHANGE will be translated to BlockValidationResult::CONSENSUS_CHANGE instead of BlockValidationResult::RECENT_CONSENSUS_CHANGE?

I don't think so. Anyone introducing `RECENT_CONSENSUS_CHANGE` should verify that it gets properly propagated up the call stack.

ryanofsky commented on src/validation.cpp:
```diff
@@ -1876,25 +1871,19 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
             std::vector<CScriptCheck> vChecks;
             bool fCacheResults = fJustCheck; /* Don't cache results if we're actually connecting blocks (still consult the cache, though) */
-            if (!CheckInputs(tx, state, view, fScriptChecks, flags, fCacheResults, fCacheResults, txdata[i], nScriptCheckThreads ? &vChecks : nullptr)) {
-                if (state.GetReason() == ValidationInvalidReason::TX_NOT_STANDARD) {
```
In commit "[validation] Add CValidationState subclasses" (b31987c45045cc26ca7867d9d8aa645f5d36b5e6)

Similar questions here as for the CheckTxInputs check above. What's the explanation for why there's no change in behavior here? Would CheckInputs never set reasons other than CONSENSUS and TX_NOT_STANDARD?

And is it a problem that the recent_consensus code will be translated to just consensus?

jnewbery replied:
> Would CheckInputs never set reasons other than CONSENSUS and TX_NOT_STANDARD?

Exactly correct.

> And is it a problem that the recent_consensus code will be translated to just consensus?

As above. Any PR that introduces recent_consensus should include tests to make sure it does what it's supposed to.

ryanofsky reviewed (commented):
Started review (will update list below with progress).
- [X] b31987c45045cc26ca7867d9d8aa645f5d36b5e6 [validation] Add CValidationState subclasses (1/7)
- [x] 22be379c54eb03e77803b0bbd8270082452d8fff [validation] Remove Reject Code from ValidationState (2/7)
- [x] b38f22795106501afb63823cc28aa768590ac5c4 [validation] Tidy Up ValidationResult class (3/7)
- [x] 1550f781fc0a86fd318883512a7104d018cf7c56 [validation] Remove error() calls from Invalid() calls (4/7)
- [x] c3d2a7c712481eedefe6da2cdfd5eb4f5859d16f [validation] Remove useless ret parameter from Invalid() (5/7)
- [x] 3b578fdca854d230592023baf5ef7863ead317b6 [validation] Remove unused first_invalid parameter from ProcessNewBlockHeaders() (6/7)
- [x] b7e1600e4df926da35b1664c32911d909899dd44 [validation] Remove fMissingInputs from AcceptToMemoryPool() (7/7)

ryanofsky commented on src/validation.cpp:
```diff
@@ -1799,8 +1799,8 @@ bool CChainState::ConnectBlock(const CBlock& block, BlockValidationState& state,
                 if (view.HaveCoin(COutPoint(tx->GetHash(), o))) {
-                    return state.Invalid(BlockValidationResult::CONSENSUS, error("ConnectBlock(): tried to overwrite transaction"),
-                                     "bad-txns-BIP30");
+                    LogPrintf("ConnectBlock(): tried to overwrite transaction\n");
```
In commit "[validation] Remove error() calls from Invalid() calls" (1550f781fc0a86fd318883512a7104d018cf7c56)

Is it intentional to drop the `ERROR:` prefix here and elsewhere in this PR?

https://github.com/bitcoin/bitcoin/blob/5d2ccf0ce9ca1571c650a69319fb9c1e0b626ecb/src/util/system.h#L60-L65

If so, it would be good to mention the change in behavior in the commit message, and maybe in release notes.

jnewbery replied:
I can't remember whether it was intentional or not. I've re-added the `ERROR: ` prefix so there's no change in logging behaviour.

ryanofsky reviewed (approved):
Conditional utACK b7e1600e4df926da35b1664c32911d909899dd44 if a few comments are addressed:

https://github.com/bitcoin/bitcoin/pull/15921#discussion_r291535754
https://github.com/bitcoin/bitcoin/pull/15921#discussion_r291556494
https://github.com/bitcoin/bitcoin/pull/15921#discussion_r291802797
https://github.com/bitcoin/bitcoin/pull/15921#discussion_r291804420

It would also be nice to get responses to:

https://github.com/bitcoin/bitcoin/pull/15921#discussion_r289036289
https://github.com/bitcoin/bitcoin/pull/15921#discussion_r289031186

My other comments are minor and could be ignored.

This cleanup is really nice and makes the ValidationState interface a lot nicer and more understandable.

promag commented on src/node/transaction.cpp:
```diff
@@ -30,16 +30,15 @@ TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx,
+        if (!AcceptToMemoryPool(mempool, state, std::move(tx),
                                 nullptr /* plTxnReplaced */, false /* bypass_limits */, highfee)) {
             if (state.IsInvalid()) {
+                if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
```
Good catch.

jnewbery replied:
It was intentional to maintain existing behaviour, but you're right that it looks odd and buggy to anyone reading the code now. Rather than write a comment explaining that I'm maintaining existing behaviour, I'm just going to set `err_string` in all cases.

promag commented on src/rpc/rawtransaction.cpp:
```diff
@@ -920,19 +920,19 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
+                result_0.pushKV("reject-reason", "missing-inputs");
+            }
             result_0.pushKV("reject-reason", strprintf("%s", state.GetRejectReason()));
```
Just one more case where univalue should have different modifiers.

Anyway looks like it's should be in ` else { ... }` block.

jnewbery replied:
Wow, that's an embarrassing bug. I think I just misread the `pushKVs` as early returns. Thanks for catching!

Yes, the fix here is to put this code in an `else` block. Done.

l2a5b1 reviewed (commented):
Nice work!

My feedback is not critical, but it could make sense to address it as we are tidying things up here.

l2a5b1 commented:
Not critical, but nice to have since we are tidying things up here:

In this PR the return type of `ValidationState::Invalid` method has been changed to `void`.

Perhaps you can also change the return type of `ValidationState::Error`, `TxValidationState ::Invalid`, `BlockValidationState::Invalid` methods to `void`? 

I don't think it is the responsibility of these methods to return `false` and make assumptions on behalf of the code at the call site.

jnewbery commented:
Rebased on master: https://github.com/bitcoin/bitcoin/compare/b7e1600e4df926da35b1664c32911d909899dd44..d79797b61ef986f6457235febdf5797a68683ad1

I still have to address all of Russ's feedback.

jnewbery commented:
I've removed the _[validation] Remove Reject Code from ValidationState_ commit from this PR (https://github.com/bitcoin/bitcoin/compare/d79797b61ef986f6457235febdf5797a68683ad1..47380ebccfb426d3cfbb44b901dafbbde839559b). See https://github.com/bitcoin/bitcoin/pull/15921#discussion_r305096212 for reason.

jnewbery commented on src/consensus/validation.h:
```diff
@@ -61,57 +57,57 @@ enum class ValidationInvalidReason {
+    BLOCK_MUTATED,           //!< the block's data didn't match the data committed to by the PoW
+    BLOCK_MISSING_PREV,      //!< We don't have the previous block the checked one is built on
+    BLOCK_INVALID_PREV,      //!< A block this one builds on is invalid
+    BLOCK_BAD_TIME,          //!< block timestamp was > 2 hours in the future (or our clock is bad)
```
Reverted. This PR was based on an earlier version of 15141, which I guess had the name BLOCK_BAD_TIME.

jnewbery commented on src/validation.cpp:
```diff
@@ -2980,28 +2969,32 @@ bool CheckBlock(const CBlock& block, CValidationState& state, const Consensus::P
 
     // Check transactions
-    for (const auto& tx : block.vtx)
-        if (!CheckTransaction(*tx, state, true))
```
> Would CheckTransaction never set a reason besides CONSENSUS?

Correct - CheckTransaction only ever sets `CONSENSUS`.

> Is it a problem that the recent_consensus

as above

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -152,7 +152,7 @@ class LockImpl : public Chain::Lock, public UniqueLock<CCriticalSection>
     {
         LockAssertion lock(::cs_main);
-        return AcceptToMemoryPool(::mempool, state, tx, nullptr /* missing inputs */, nullptr /* txn replaced */,
+        return AcceptToMemoryPool(::mempool, state, tx, nullptr /* missing inputs */,
```
fixed

jnewbery commented on src/consensus/validation.h:
```diff
@@ -61,84 +46,101 @@ enum class ValidationInvalidReason {
+    std::string m_debug_message;
+protected:
+    void Invalid(const std::string &reject_reason="",
+                 const std::string &debug_message="")
```
I'd like to remove the optional parameters, but there are many calls to `Invalid()` that don't set a debug message.

jnewbery commented on src/consensus/validation.h:
```diff
@@ -61,84 +46,101 @@ enum class ValidationInvalidReason {
+/** Base class for capturing information about block/transaction validation. This is subclassed
+ *  by TxValidationState and BlockValidationState for validation information on transactions
+ *  and blocks respectively. */
+class ValidationState {
```
Done. Thanks for the review!

jkczyz replied:
Sorry for the late comment. Would a template class be more suitable? It appears that the subclasses only differ by the type of a member variable. So making this a template would remove a bit of boilerplate. Then a typedef could be used to keep the subclass names if desired.

jkczyz replied:
Made the change in #17399. Happy to close the PR if this isn't desirable for some reason.

jnewbery commented:
I believe I've now addressed all comments. This should be ready for review.

@l2a5b1
> Perhaps you can also change the return type of ValidationState::Error, TxValidationState ::Invalid, BlockValidationState::Invalid methods to void?

Seems like a reasonable change, but I'd rather not expand this PR even more. It's already making a lot of changes.

jnewbery commented:
Fixed the bad rebase in validation_block_tests.cpp.

jnewbery commented:
rebased

jnewbery commented:
Rebased on master. This is now based on #17004. Please review that first.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/15921#pullrequestreview-247365826

> Conditional utACK [b7e1600](https://github.com/bitcoin/bitcoin/commit/b7e1600e4df926da35b1664c32911d909899dd44) if a few comments are addressed

It's been a while, but this is next on my review list after 16341 (giant achow keyman pr), so I didn't forget about this. It may help to add "**Based on  #15437 and #17004**" to the top of the PR description to make this seem less big and show other reviewers where to look first.

ryanofsky commented on src/consensus/validation.h:
```diff
@@ -15,7 +15,7 @@
 /** A "reason" why something was invalid, suitable for determining whether the
   * provider of the object should be banned/ignored/disconnected/etc.
```
In commit "[validation] Add CValidationState subclasses" (907ef02667b6c224df1457b67b7d7950cd58494f)

A previous version of this commit changed "something" to "a transaction" and "object" to "tx". Would be nice to keep these changes.

jnewbery replied:
fixed

ryanofsky commented on src/validation.cpp:
```diff
@@ -1547,7 +1546,7 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
+                    return state.Invalid(TxValidationResult::TX_NOT_STANDARD, false, strprintf("non-mandatory-script-verify-flag (%s)", ScriptErrorString(check.GetScriptError())));
             }
             // MANDATORY flag failures correspond to
             // ValidationInvalidReason::CONSENSUS. Because CONSENSUS
```
In commit "[validation] Add CValidationState subclasses" (907ef02667b6c224df1457b67b7d7950cd58494f)

Could s/ValidationInvalidReason/TxValidationResult/ here like previous versions of this commit

jnewbery replied:
fixed

jnewbery commented:
Thanks for the review @ryanofsky . I've addressed both of your comments.

jnewbery commented:
rebased. Still depends on #17004, so please review that PR first.

ryanofsky commented on src/validation.cpp:
```diff
@@ -2136,25 +2131,19 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
-                              state.GetRejectReason(), state.GetDebugMessage());
-                }
+            TxValidationState tx_state;
+            if (!CheckInputs(tx, tx_state, view, flags, fCacheResults, fCacheResults, txdata[i], nScriptCheckThreads ? &vChecks : nullptr)) {
```
In commit "[validation] Add CValidationState subclasses" (b885aec88a627363bf6771c941e855fa735145c6)

Is it intentional to drop the `fScriptChecks && ` condition above?

jnewbery replied:
eeek. Good catch. This was a bad rebase (and would have broken `-assumevalid` if it'd been merged).

jnewbery commented:
Fixed the bad rebase pointed out by @ryanofsky (https://github.com/bitcoin/bitcoin/pull/15921#discussion_r335106673): https://github.com/bitcoin/bitcoin/compare/6fec91c1212c41698d1c9464379d17627cccf42a..b68a33829827c51048784a6f83f8acea3351d3c0

jnewbery commented:
rebased: https://github.com/bitcoin/bitcoin/compare/b68a33829827c51048784a6f83f8acea3351d3c0..735e86fac20adee2c42e34ee382aaea95fbdbbb3

ryanofsky reviewed (approved):
Code review ACK 735e86fac20adee2c42e34ee382aaea95fbdbbb3. There was a pretty big rebase since my last ACK. I rereviewed the first state subclasses commit and dug into ConnectBlock check calls more to verify no behavior should be changing. For the later commits I rebased and compared the changes against the previous PR. Only changes were various things discussed in comments: bugfixes, adding back log ERROR prefixes, and err_string setting in ATMP.

jnewbery commented:
Thanks for the rereview @ryanofsky !

jnewbery commented:
rebased

jnewbery committed:
[validation] Add CValidationState subclasses

Split CValidationState into TxValidationState and BlockValidationState
to store validation results for transactions and blocks respectively.

jnewbery commented:
Rebased on master. Also tidied up a few small details (no functional changes)

ryanofsky reviewed (approved):
Code review ACK 9ec6b4fb30142c182212fb24a581774a18fca00c. Not substantial changes since last review, mainly rebase. Also new assert, RECENT_CONSENSUS enums, some more `TX_` and `BLOCK_` enum prefixes, and comments in the first commit. And more removed error() calls in the third commit, I think restoring a removed ERROR: log prefix.

jnewbery commented:
The linter caught a log that wasn't terminated with a newline. Pushed a fix.

ryanofsky reviewed (approved):
Code review ACK 72d303bf60a87a6ee559ca06627fff4f2d3457a0. Changes since last review: fixed log newline and increased linter happiness

mzumsande commented:
The fuzzing harness `src/test/fuzz/transaction.cpp` added in #17076 uses `CValidationState` and needs an update.

jnewbery commented:
> The fuzzing harness src/test/fuzz/transaction.cpp added in #17076 uses CValidationState and needs an update.

Thanks @mzumsande . Now fixed. I've also rebased on master.

ryanofsky reviewed (approved):
Code review ACK dcf5003ca0d8927f3e2457a32394a70ed683f01e. Changes since last review: rebase after fCheckDuplicateInputs removal and fuzz test compile fix

laanwj commented:
ACK dcf5003ca0d8927f3e2457a32394a70ed683f01e

mzumsande commented on src/consensus/validation.h:
```diff
@@ -48,64 +43,67 @@ enum class ValidationInvalidReason {
     }
+public:
+    // We use ValidationState polymorphically. Have a virtual destructor
+    virtual ~ValidationState() {}
```
Your old [answer](https://github.com/bitcoin/bitcoin/pull/15921#discussion_r305146828) to  to an old [comment](https://github.com/bitcoin/bitcoin/pull/15921#discussion_r299230821) indicates that at some point you wanted to make the base class `ValidationState` abstract. However, the class has a (non-pure) virtual destructor and no pure virtual functions, so instantiation of `ValidationState` should still be possible. Is this working as intended?

fjahr commented on src/consensus/tx_check.cpp:
```diff
@@ -7,28 +7,28 @@
 #include <consensus/validation.h>
 
-bool CheckTransaction(const CTransaction& tx, CValidationState& state)
+bool CheckTransaction(const CTransaction& tx, TxValidationState &state)
```
nit: the `&` was moved from the type to the variable name. For consistency I would keep it at end of the type. That's also how it is still declared in the header file.

fjahr reviewed (approved):
Code review ACK dcf5003. Also built and ran tests locally.

I agree with @JeremyRubin that `Invalid` should return void ideally. Also, have you thought about moving the punishment logic into its own class and file? It's not much of a difference in terms of LoC but reducing `net_processing.cpp` in size would not be a bad thing and it feels like that logic could deserve its own place due to its importance.

jnewbery commented:
Thanks for the reviews @mzumsande and @fjahr . I've added a couple of fixup commits that address your comments. Let me know if you like them and I'll squash into the appropriate commits.

@mzumsande - I had to add a consensus/validation.cpp file to define the `ValidationState` destructor. Let me know if you know of a better way to do this.

mzumsande commented:
For me, `inline ValidationState::~ValidationState() {};` in the header works as well to make the class abstract (the `inline` prevents `multiple definition` linker errors). I am not sure about the cause for the MSVC/AppVeyor of the current implementation.

jnewbery commented:
> For me, inline ValidationState::~ValidationState() {}; in the header works 

Thanks. Much better!

fjahr commented:
> Thanks for the reviews @mzumsande and @fjahr . I've added a couple of fixup commits that address your comments. Let me know if you like them and I'll squash into the appropriate commits.
> 
> @mzumsande - I had to add a consensus/validation.cpp file to define the `ValidationState` destructor. Let me know if you know of a better way to do this.

Looks good to me!

jnewbery commented:
squashed fixup commits.

ryanofsky reviewed (approved):
Code review ACK 3004d5a12d09d94bfc4dee2a8e8f2291996a4aaf. Just whitespace change and pure virtual destructor added since last review.

amitiuttarwar commented:
code review ACK 3004d5a12d09d94bfc4dee2a8e8f2291996a4aaf. Built & ran tests locally.

fjahr commented:
Code review ACK 3004d5a. Only nit style change and pure virtual destructor added since my last review.

jonatack commented:
There oughta be a rule about merging review club PRs right before the meeting :)

laanwj commented:
@jonatack whoops …
nah, reviewing doesn't necessarily stop after merging :smile:

MarcoFalke commented:
>  nah, reviewing doesn't necessarily stop after merging

Agree. The only downside is that review comments (with the reviewer's name) can not be included in the merge commit.

jonatack commented:
Could possibly use a `review club` tag :label:

jonatack reviewed (commented):
Built at 3004d5a12d09d94bfc4dee2a8e8f2291996a4aaf, ran tests, running bitcoind tailing the log.

Code review progress:
- [x] [validation] Add CValidationState subclasses
- [x] [validation] Tidy Up ValidationResult class 
- [x] [validation] Remove error() calls from Invalid() calls
- [x] [validation] Remove useless ret parameter from Invalid()
- [x] [validation] Remove unused first_invalid parameter from ProcessNewBlo…
- [ ] [validation] Remove fMissingInputs from AcceptToMemoryPool()

Interesting reviewer tip in the PR description; [result here](https://gist.github.com/jonatack/3d268e1e8fc8f095925bd2a523bf9306) (without the right colors).

laanwj commented:
> Could possibly use a review club tag label

Done

jonatack commented:
Perhaps helpful context for future readers/reviewers, here are two comments motivating a27a2957ed9afbe5a96caa5f0f4cbec730d27460 (CValidationState subclasses) from discussion in #15141:

- sipa: https://github.com/bitcoin/bitcoin/pull/15141#issuecomment-461933821
- theBlueMatt: https://github.com/bitcoin/bitcoin/pull/15141#discussion_r273667031

jonatack commented:
> > Could possibly use a review club tag label
> 
> Done

Thanks!

practicalswift commented:
Reviewers of this PR are encouraged to review PR #17624 ("net: Fix an uninitialized read in ProcessMessage(…, "tx", …) when receiving a transaction we already have") which fixes a quite serious bug introduced in this PR. Luckily it was caught before being part of any release :)
